src/Shared/Buffers/RefPooledArrayBufferWriter.cs

```diff
@@ -0,0 +1,266 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+// Copied from https://github.com/dotnet/corefx/blob/b0751dcd4a419ba6731dcaa7d240a8a1946c934c/src/System.Text.Json/src/System/Text/Json/Serialization/ArrayBufferWriter.cs
+
+using System.Diagnostics;
+using System.Runtime.CompilerServices;
+
+namespace System.Buffers;
+
+/// <summary>
+/// A high-performance struct-based IBufferWriter&lt;byte&gt; implementation that uses ArrayPool for allocations.
+/// Designed for zero-allocation scenarios when used with generic methods via `allows ref struct` constraint.
+/// </summary>
+internal ref struct RefPooledArrayBufferWriter<T> : IBufferWriter<T>, IDisposable
+{
+    private T[] _rentedBuffer;
+    private Span<T> _buffer;
+    private int _index;
+
+    private const int MinimumBufferSize = 256;
+
+    /// <summary>
+    /// Initializes the <see cref="RefPooledArrayBufferWriter{T}"/> with initial buffer. 
+    /// </summary>
+    /// <param name="initialBuffer">The initial buffer to start writer with.</param>
+    public RefPooledArrayBufferWriter(Span<T> initialBuffer)
+    {
+        // no rented buffer initially - only if we need to grow over the limits of initialBuffer
+        _rentedBuffer = null!;
+
+        _buffer = initialBuffer;
+        _index = 0;
+    }
+
+    /// <summary>
+    /// Clears the buffer contents and returns the rented array to the ArrayPool.
+    /// This must be called to properly clean up resources.
+    /// </summary>
+    public void Dispose()
+    {
+        // to avoid `bool isDisposed` field, we can use negative index as disposed marker
+        _index = -1;
+
+        if (_rentedBuffer is not null)
+        {
+            // Optionally clear the buffer before returning to pool (security)
+            // Uncomment if needed for sensitive data:
+            // _buffer.AsSpan(0, _index).Clear();
+
+            ArrayPool<T>.Shared.Return(_rentedBuffer, clearArray: false);
+            _buffer = null!;
```

Comment:
No reason to leave the dead code around or to be explicit about `clearArray` when not clearing is the well-known default. But if we did want to reduce risk, we could create an ArrayPool using `ArrayPool<byte>.Create()`. that's used exclusively for DataProtection rather than use the shared one. This would also give us more control over the max size of pooled arrays and the maximum number of arrays to pool.

It's also unclear what nulling out the buffer does. Presumably the stack frame with the ref struct will be popped shortly after this will be disposed.

```suggestion
            ArrayPool<T>.Shared.Return(_rentedBuffer);
```